service/history/historyEngine.go

```diff
@@ -2713,9 +2750,69 @@ func (e *historyEngineImpl) ReapplyEvents(
 				reapplyEvents,
 			); err != nil {
 				e.logger.Error("failed to re-apply stale events", tag.Error(err))
-				return nil, &workflow.InternalServiceError{Message: "unable to re-apply stale events"}
+				return emptyUpdateAction, &workflow.InternalServiceError{Message: "unable to re-apply stale events"}
 			}
 
 			return postActions, nil
 		})
 }
+
+func (e *historyEngineImpl) loadWorkflow(
+	ctx ctx.Context,
+	domainID string,
+	workflowID string,
+	runID string,
+) (workflowContext, error) {
+
+	context, release, err := e.historyCache.getOrCreateWorkflowExecution(
+		ctx,
+		domainID,
+		workflow.WorkflowExecution{
+			WorkflowId: common.StringPtr(workflowID),
+			RunId:      common.StringPtr(runID),
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	mutableState, err := context.loadWorkflowExecution()
+	if err != nil {
+		release(err)
+		return nil, err
+	}
+
+	return newWorkflowContext(context, release, mutableState), nil
+}
+
+func (e *historyEngineImpl) loadCurrentWorkflow(
+	ctx ctx.Context,
+	domainID string,
+	workflowID string,
+	runID string,
+) (workflowContext, error) {
+
+	workflowContext, err := e.loadWorkflow(ctx, domainID, workflowID, runID)
+	if err != nil {
+		return nil, err
+	}
+
+	if runID == "" {
+		for attempt := 0; attempt < conditionalRetryCount && !workflowContext.getMutableState().IsWorkflowExecutionRunning(); attempt++ {
+
+			prevRunID := workflowContext.getRunID()
+			workflowContext.getReleaseFn()(nil)
+
+			workflowContext, err = e.loadWorkflow(ctx, domainID, workflowID, runID)
+			if err != nil {
+				return nil, err
+			}
+
+			if workflowContext.getRunID() == prevRunID {
+				workflowContext.getReleaseFn()(ErrWorkflowCompleted)
+				return nil, ErrWorkflowCompleted
+			}
+		}
+	}
+	return workflowContext, err
```

Comment:
i would write it is this way
```go
if runID != "" {
	return e.loadWorkflow(ctx, domainID, workflowID, runID)
}

// loop to check & get the current workflow
for attempt := 0; attempt < conditionalRetryCount; attempt++ {
	
	workflowContext, err := e.loadWorkflow(ctx, domainID, workflowID, "")
	if err != nil {
		return nil, err
	}

	if workflowContext.getMutableState().IsWorkflowExecutionRunning() {
		return workflowContext, nil
	}

	// workflow not running, need to check current record
	resp, err := r.shard.GetExecutionManager().GetCurrentExecution(
		&persistence.GetCurrentExecutionRequest{
			DomainID:   domainID,
			WorkflowID: workflowID,
		},
	)
	if err != nil {
		workflowContext.getReleaseFn()(err)
		return nil, err
	}

	if resp.RunID == workflowContext.GetRunID() {
		return workflowContext, nil
	}
	workflowContext.getReleaseFn()(nil)
}
```

Reply:
we need let caller make the judgement whether to return `ErrWorkflowCompleted` or not, not within this function

Reply:
Accepted and updated.